File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> temp;
        temp.push_back(nums[0]);
        for(int i = 1; i < nums.size(); i++) {
            if(nums[i] > temp.back()) temp.push_back(nums[i]);
            else {
                int idx = lower_bound(temp.begin(), temp.end(), nums[i]) - temp.begin();
                temp[idx] = nums[i];
            }
        }

        return temp.size();
    }
};

////// DP ///////

// class Solution {
// public:
//     int lengthOfLIS(vector<int>& nums) {
//         int n = nums.size();
//         vector<int> dp(n, 1);
//         for(int i = 0; i < n; i++){
//             for(int j = 0; j < i; j++){
//                 if(nums[i] > nums[j]){
//                     dp[i] = max(dp[i], dp[j] + 1);
//                 }
//             }
//         }
//         return *max_element(dp.begin(), dp.end());
//     }
// };

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    private int calculateLength(int idx, int prevIdx, int[] nums, int[][] dp) {
        if (idx >= nums.length) return 0;
        if (dp[idx][prevIdx + 1] != -1) return dp[idx][prevIdx + 1];
        
        int take = 0;
        if(prevIdx == -1 || nums[idx] > nums[prevIdx]) {
            take = 1 + calculateLength(idx + 1, idx, nums, dp);
        }
        
        int skip = calculateLength(idx + 1, prevIdx, nums, dp);

        dp[idx][prevIdx + 1] = Math.max(take, skip);
        return dp[idx][prevIdx + 1];
    }
    
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n][n];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return calculateLength(0, -1, nums, dp);
    }
}